source/lac/cuda_vector.cu

```diff
@@ -0,0 +1,1006 @@
+#include <deal.II/lac/read_write_vector.h>
+#include <deal.II/base/exceptions.h>
+#include <cmath>
+
+#ifdef DEAL_II_WITH_CUDA
+
+DEAL_II_NAMESPACE_OPEN
+
+#define BKSIZE_ELEMWISE_OP 512
+#define CHUNKSIZE_ELEMWISE_OP 8
+#define VR_BKSIZE 512
+#define VR_CHUNK_SIZE 8
+
+namespace LinearAlgebra
+{
+  namespace CUDAWrappers
+  {
+    namespace internal
+    {
+      template <typename Number>
+      __global__ void vec_scale(Number                                   *val,
+                                const Number                              a,
+                                const typename Vector<Number>::size_type  N)
+      {
+        const typename Vector<Number>::size_type idx_base = threadIdx.x +
+                                                            blockIdx.x *
+                                                            (blockDim.x*CHUNKSIZE_ELEMWISE_OP);
+        for (unsigned int i=0; i<CHUNKSIZE_ELEMWISE_OP; ++i)
+          {
+            const typename Vector<Number>::size_type idx = idx_base +
+                                                           i*BKSIZE_ELEMWISE_OP;
+            if (idx<N)
+              val[idx] *= a;
+          }
+      }
+
+
+
+      struct Binop_Addition
+      {
+        template <typename Number>
+        __device__ static inline Number operation(const Number a,
+                                                  const Number b)
+        {
+          return a+b;
+        }
+      };
+
+
+
+      struct Binop_Subtraction
+      {
+        template <typename Number>
+        __device__ static inline Number operation(const Number a,
+                                                  const Number b)
+        {
+          return a-b;
+        }
+      };
+
+
+
+      template <typename Number, typename Binop>
+      __global__ void vector_bin_op(Number                                   *v1,
+                                    Number                                   *v2,
+                                    const typename Vector<Number>::size_type  N)
+      {
+        const typename Vector<Number>::size_type idx_base = threadIdx.x +
+                                                            blockIdx.x *
+                                                            (blockDim.x*CHUNKSIZE_ELEMWISE_OP);
+        for (unsigned int i=0; i<CHUNKSIZE_ELEMWISE_OP; ++i)
+          {
+            const typename Vector<Number>::size_type idx = idx_base +
+                                                           i*BKSIZE_ELEMWISE_OP;
+            if (idx<N)
+              v1[idx] = Binop::operation(v1[idx],v2[idx]);
+          }
+      }
+
+
+      template <typename Number>
+      struct L1Norm
+      {
+        __device__ static Number reduction_op(const Number a, const Number b)
+        {
+          return std::abs(a) + std::abs(b);
+        }
+
+        __device__ static void atomic_op(Number *dst, const Number a)
+        {
+          *dst = std::abs(*dst) + std::abs(a);
```

Comment:
Yeah, we definitely need atomics in the last part of the reduction. We do reduction in CUDA by
0. Perform any element-wise operations, such as absolute-value in the case of 1-norm.
1. Reduce within thread blocks with the given binary operation, in which case it is done in shared memory with synchronized writes, thus no atomics are needed.
2. Now we have one value per thread block which we need to reduce to a single value, but we cannot synchronize updates of threads in different blocks, so we have to use atomics.

This is the reason why I used two different operations in the input struct. Really, there should be the third element-wise operation initially, to separate out the abs from the additions. The way it is done now is not wrong, but all the intermediate abs's are not needed. The whole idea with unifying the different reductions seems a bit messy now, but I there is quite a lot of code which would be duplicated otherwise. Let's think about this a bit more.